docs/cardano-monads.rst

```diff
@@ -0,0 +1,827 @@
+exceptions). Mixing them is unhelpful.
+
+Now, let's consider the global ``Env`` type. The issue with it is associated
+boilerplate of ``Has`` classes. In case we want to extend the environment to
+introduce local effects, there's a high cost in boilerplate. This can be already
+observed in our codebase.
+
+Point-by-point rundown:
+
+* Flexibility: LOW. Extension of environment requires declaring a new type,
+  writing ``Has`` instances for it, and potentially monad instances (although the
+  approach recommends avoiding monad classes in favor of ``Has`` classes, not
+  everything is under our control and there are classes by external libraries).
+  There's no way to partially run effects -- one needs to supply the entire
+  ``ReaderT`` environment up front.
+
+* Extensibility: MODERATE. Defining new effects boils down to declaring a method
+  record and a ``Has`` class for it. Then, for each field in the method record we
+  must define a function that extracts this method using the ``Has`` class and
+  runs it. Boilerplate-heavy, yes, but at least effects don't need to know about
+  each other.
+
+* Ease of use: MODERATE. Beginners can quickly grasp the basic concepts, but
+  there are pitfalls when it comes to lifting method records (see section on
+  ``SendActions``).
+
+* Compile-time performance: SUPERB. There's nothing to it, GHC can easily handle
+  this style of code.
+
+* Run-time performance: MODERATE. A flat ``ReaderT`` layer is good, but method
+  records are stored within the context (rather than passed via instance
+  search). Because of this, the compiler can't assume that the method record is
+  coherent (it's not, you can modify it with ``local``), and it limits inlining.
+  Instance specialization can't be performed because there are no instances to
+  speak of.
+
+* Run-time configurability: SUPERB. You can put just about anything into these
+  manually defined method records, modify them at will, etc. However, we need to
+  figure out a good story for method records that depend on other method
+  records, because it's the same pitfall as with ``SendActions``: how can we
+  ensure that if we change the something (say, a logging method) with ``local``,
+  other method records that use logging will be updated accordingly?
+
+* Predictability: LOW/HIGH. Depends on whether we have a ``newtype`` over
+  ``ReaderT``.
+
+Verdict: the approach is viable but has its costs.
+
+
+Dictionary-Passing Style
+~~~~~~~~~~~~~~~~~~~~~~~~
+
+Similar to previous approach and one was widely used by the networking team.
+
+As we understand ideas behind it:
+
+- Single base monad
+
+  - Similarly to ``ReaderT IO`` approach
+
+- Differentiating implementations all stored in records
+
+  - No type classes and instances like ``instance MyClass (ReaderT Ctx IO)``
+  - Each function accepts a single **explicit** parameter of type ``Ctx m``
+  - This type ``Ctx`` contains all needed methods
+
+Point-by-point rundown:
+
+* Flexibility: LOW. (Same as above).
```

Comment:
It's not the same as above. Quoting above

```
Flexibility: LOW. Extension of environment requires declaring a new type,
writing ``Has`` instances for it, and potentially monad instances (although the
approach recommends avoiding monad classes in favor of ``Has`` classes, not
everything is under our control and there are classes by external libraries).
There's no way to partially run effects -- one needs to supply the entire
`ReaderT` environment up front.
```

Extending in dictionary-passing style doesn't require defining any new instances, nor in many cases defining any new types (perhaps defining a new record field). Extension of the environment is rather localized. There is not necessarily one global environment that must be constructed (there's no `ReaderT r`). There are local environments for local features, which can be locally modified without necessitating a global change.

Reply:
The assumption was that all method dictionaries are passed in a single `Ctx` type with `Has` instances for it, otherwise we have to pass each dictionary as a separate parameter, so we need to declare new `Has` instances to change `Ctx` into `(LocalCtx, Ctx)`. However, I agree that there are ways to work around that (I described the workaround in the "hybrid approach" section), and the low rating is unfair. I will change to "moderate" in both cases.